File: client/test/first-content-candidates.test.ts
import { describe, expect, it } from "vitest";

import achievements from "../../balance/testdata/first-content/achievements-v1.json?raw";
import categories from "../../balance/testdata/first-content/categories-v1.json?raw";
import commons from "../../balance/commons/phase0.json?raw";
import doctrines from "../../balance/testdata/first-content/doctrines-v1.json?raw";
import economy from "../../balance/testdata/first-content/economy-v3.json?raw";
import factions from "../../balance/factions/phase0.json?raw";
import fiscal from "../../balance/testdata/first-content/fiscal-v1.json?raw";
import guilds from "../../balance/guilds/phase0.json?raw";
import meters from "../../balance/testdata/first-content/meters-v1.json?raw";
import minigameAPI from "../../balance/testdata/minigame-api-candidate-v1.json?raw";
import minigames from "../../testdata/minigame/pitch-v3.json?raw";
import pets from "../../balance/testdata/first-content/pets-v2.json?raw";
import pitch from "../../balance/testdata/pitch-v1.json?raw";
import prestige from "../../balance/prestige/phase0.json?raw";
import routes from "../../balance/testdata/permits-t3-gate-candidate-v1.json?raw";
import soul from "../../balance/testdata/first-content/soul-v1.json?raw";
import manifest from "../../planning/first-content-epoch/promotion-manifest.candidate.v1.json";
import { loadReplayCatalogBundle, type ReplayArtifacts } from "../src/replay";

const artifacts: ReplayArtifacts = {
  achievements,
  categories,
  commons,
  doctrines,
  economy,
  factions,
  fiscal,
  guilds,
  meters,
  minigame_api: minigameAPI,
  minigames,
  pets,
  pitch,
  prestige,
  routes,
  soul,
};

describe("first-content candidate artifacts", () => {
  it("loads the same complete sixteen-artifact bundle as the Go runtime", async () => {
    const hash = await constantsHashArtifacts(artifacts);
    expect(hash).toBe(manifest.constants_hash);
    const bundle = await loadReplayCatalogBundle(hash, artifacts);
    expect(bundle.constantsHash).toBe(hash);
    expect(bundle.meters?.meters).toHaveLength(11);
    expect(bundle.achievements?.definitions).toHaveLength(12);
    expect(bundle.doctrines?.transitions).toHaveLength(1);
    expect(bundle.minigames?.minigames).toHaveLength(1);
    expect(bundle.pets?.schema_version).toBe(2);
    expect(bundle.fiscal?.unlockRows.map((row) => row.unlockId)).toEqual(["minigame.pitch", "unlock.arcade"]);
    expect(bundle.soul?.recovery_activities).toHaveLength(3);
    expect(bundle.pitch?.metric_cards).toHaveLength(12);
    expect(bundle.minigameAPI?.tenants).toHaveLength(1);
  });

  it("rejects the pre-permits leaderboard gate set", async () => {
    const staleCategories = JSON.stringify({ ...JSON.parse(categories), full_gate_set: ["gate.t2_to_t3", "gate.t4_to_t5", "gate.t7_to_t8"] });
    const stale = { ...artifacts, categories: staleCategories };
    await expect(loadReplayCatalogBundle(await constantsHashArtifacts(stale), stale)).rejects.toThrow();
  });
});

async function constantsHashArtifacts(values: ReplayArtifacts): Promise<string> {
  const encoder = new TextEncoder();
  const chunks: Uint8Array[] = [];
  for (const name of Object.keys(values).sort() as Array<keyof ReplayArtifacts>) {
    const nameBytes = encoder.encode(name);
    const data = encoder.encode(values[name]);
    chunks.push(frame(nameBytes.length), nameBytes, frame(data.length), data);
  }
  const input = new Uint8Array(chunks.reduce((sum, value) => sum + value.length, 0));
  let offset = 0;
  for (const chunk of chunks) {
    input.set(chunk, offset);
    offset += chunk.length;
  }
  const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", input));
  return `sha256:${[...digest].map((value) => value.toString(16).padStart(2, "0")).join("")}`;
}

function frame(value: number): Uint8Array {
  const result = new Uint8Array(8);
  new DataView(result.buffer).setBigUint64(0, BigInt(value), false);
  return result;
}

File: server/replaycatalog/first_content_candidates_test.go
package replaycatalog

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"cloud-clicker/server/epochseed"
	"cloud-clicker/server/save"
)

func TestT0T1MintPromotesRatifiedBytesAndPreservesFirstContent(t *testing.T) {
	root := filepath.Join("..", "..")
	manifest := t0T1CandidateManifest(t)
	bundle, err := epochseed.Load(root)
	if err != nil {
		t.Fatal(err)
	}
	if bundle.Seed.CurrentEpochID != 7 || len(bundle.Artifacts) != 18 || bundle.Hash != manifest.ConstantsHash ||
		!epochseed.Accepts(epochseed.Current(bundle.Seed), manifest.ConstantsHash) {
		t.Fatalf("mint identity epoch=%d artifacts=%d hash=%s", bundle.Seed.CurrentEpochID, len(bundle.Artifacts), bundle.Hash)
	}
	for _, row := range manifest.Artifacts {
		source, sourceErr := os.ReadFile(filepath.Join(root, row.SourcePath))
		production, productionErr := os.ReadFile(filepath.Join(root, row.ProductionPath))
		if sourceErr != nil || productionErr != nil || !bytes.Equal(source, production) {
			t.Fatalf("%s source/production mismatch source_err=%v production_err=%v", row.Name, sourceErr, productionErr)
		}
	}
	loaded, err := Load(bundle.Hash, bundle.Artifacts)
	if err != nil {
		t.Fatalf("load minted bundle: %v", err)
	}
	if loaded.Opportunities == nil || loaded.Relevance == nil {
		t.Fatal("epoch-7 bundle omitted its new artifact owners")
	}

	epoch6Artifacts := firstContentCandidateArtifacts(t)
	epoch6Hash, err := save.ConstantsHashArtifacts(epoch6Artifacts)
	if err != nil || epoch6Hash != bundle.Seed.Epochs[5].AcceptedHashes[0] {
		t.Fatalf("epoch-6 fixture hash=%s err=%v", epoch6Hash, err)
	}
	if _, err := Load(epoch6Hash, epoch6Artifacts); err != nil {
		t.Fatalf("historical epoch-6 bundle no longer loads: %v", err)
	}
}

func TestFirstContentCandidateBundleLoadsFromLiteralArtifacts(t *testing.T) {
	artifacts := firstContentCandidateArtifacts(t)
	hash, err := save.ConstantsHashArtifacts(artifacts)
	if err != nil {
		t.Fatal(err)
	}
	manifest := firstContentCandidateManifest(t)
	if hash != manifest.ConstantsHash {
		t.Fatalf("candidate constants hash=%s manifest=%s", hash, manifest.ConstantsHash)
	}
	t.Logf("first-content candidate constants hash: %s", hash)
	bundle, err := Load(hash, artifacts)
	if err != nil {
		t.Fatalf("load first-content candidate %s: %v", hash, err)
	}
	if bundle.Meters == nil || bundle.Achievements == nil || bundle.Doctrines == nil || bundle.Minigames == nil ||
		bundle.Pets == nil || bundle.Fiscal == nil || bundle.Soul == nil || bundle.Pitch == nil || bundle.MinigameAPI == nil {
		t.Fatalf("candidate bundle omitted an activated catalog: %+v", bundle)
	}
}

func TestFirstContentCandidateCategorySetIsLoadBearing(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("..", "..", "balance", "testdata", "first-content", "categories-v1.json"))
	if err != nil {
		t.Fatal(err)
	}
	var candidate map[string]any
	if err := json.Unmarshal(data, &candidate); err != nil {
		t.Fatal(err)
	}
	candidate["full_gate_set"] = []string{"gate.t2_to_t3", "gate.t4_to_t5", "gate.t7_to_t8"}
	stale, err := json.Marshal(candidate)
	if err != nil {
		t.Fatal(err)
	}
	artifacts := firstContentCandidateArtifacts(t)
	artifacts["categories"] = stale
	hash, err := save.ConstantsHashArtifacts(artifacts)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := Load(hash, artifacts); err == nil {
		t.Fatal("candidate bundle accepted the pre-permits leaderboard gate set")
	}
}

func TestFirstContentCandidateRejectsMeterResourceCollision(t *testing.T) {
	artifacts := firstContentCandidateArtifacts(t)
	var economy map[string]any
	if err := json.Unmarshal(artifacts["economy"], &economy); err != nil {
		t.Fatal(err)
	}
	economy["resources"] = append(economy["resources"].([]any), map[string]any{
		"id": "trust.users.standing", "scope": "company", "numeric_kind": "decimal", "initial": "0", "minimum": "0",
		"hardcap": map[string]any{"amount": "1e2", "reason_key": "resource.trust_users_standing.cap.fixture"},
	})
	var err error
	artifacts["economy"], err = json.Marshal(economy)
	if err != nil {
		t.Fatal(err)
	}
	hash, err := save.ConstantsHashArtifacts(artifacts)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := Load(hash, artifacts); err == nil {
		t.Fatal("Go replay loader accepted an economy resource that collides with a meter ID")
	}
}

func firstContentCandidateArtifacts(t *testing.T) map[string][]byte {
	t.Helper()
	manifest := firstContentCandidateManifest(t)
	root := filepath.Join("..", "..")
	artifacts := make(map[string][]byte, len(manifest.Artifacts))
	prior := ""
	for _, row := range manifest.Artifacts {
		if row.Name <= prior || row.Name == "" || row.SourcePath == "" || row.ProductionPath == "" || row.SchemaVersion < 1 || row.ContentGate == "" || row.ConsumedVerdict == "" {
			t.Fatalf("invalid promotion manifest row for %q", row.Name)
		}
		data, err := os.ReadFile(filepath.Join(root, row.SourcePath))
		if err != nil {
			t.Fatalf("read %s: %v", row.Name, err)
		}
		actual := fmt.Sprintf("%x", sha256.Sum256(data))
		if actual != row.SHA256 {
			t.Fatalf("%s SHA-256=%s manifest=%s", row.Name, actual, row.SHA256)
		}
		artifacts[row.Name] = data
		prior = row.Name
	}
	return artifacts
}

type firstContentManifest struct {
	SchemaVersion int                       `json:"schema_version"`
	Status        string                    `json:"status"`
	ConstantsHash string                    `json:"constants_hash"`
	Artifacts     []firstContentManifestRow `json:"artifacts"`
}

type firstContentManifestRow struct {
	Name            string `json:"name"`
	SchemaVersion   int    `json:"schema_version"`
	SourcePath      string `json:"source_path"`
	ProductionPath  string `json:"production_path"`
	SHA256          string `json:"sha256"`
	ContentGate     string `json:"content_gate"`
	ConsumedVerdict string `json:"consumed_verdict"`
}

func firstContentCandidateManifest(t *testing.T) firstContentManifest {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "..", "planning", "first-content-epoch", "promotion-manifest.candidate.v1.json"))
	if err != nil {
		t.Fatal(err)
	}
	var manifest firstContentManifest
	if err := json.Unmarshal(data, &manifest); err != nil || manifest.SchemaVersion != 1 || manifest.Status != "ratified" || len(manifest.Artifacts) != 16 {
		t.Fatalf("invalid first-content promotion manifest: %v", err)
	}
	return manifest
}

func t0T1CandidateManifest(t *testing.T) firstContentManifest {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "..", "planning", "t0-t1-content", "promotion-manifest.candidate.v1.json"))
	if err != nil {
		t.Fatal(err)
	}
	var manifest firstContentManifest
	if err := json.Unmarshal(data, &manifest); err != nil || manifest.SchemaVersion != 1 || manifest.Status != "ratified" || len(manifest.Artifacts) != 18 {
		t.Fatalf("invalid T0-T1 promotion manifest: %v", err)
	}
	return manifest
}
